Sming/Components/OtaUpgrade/OtaUpgradeStream.h

```diff
@@ -0,0 +1,181 @@
+ * 	- 4 Byte Length
+ * 	- ROM Content (Length bytes)
+ * 64 Byte cryptographic signature (over everything since beginning of file)
+ * Optional extensions may follow in the future for
+ * - updating rboot
+ * - updating RF parameter blob
+ * - update/clear system parameters area
+ * - Files or file system updates
+ */
+class OtaUpgradeStream : public ReadWriteStream
+{
+public:
+	OtaUpgradeStream();
+	~OtaUpgradeStream();
+
+	enum ErrorCode {
+		NoError,
+		InvalidFormatError,
+		UnsupportedDataError,
+		DecryptionError,
+		NoRomFoundError,
+		RomTooLargeError,
+		DowngradeError,
+		VerificationError,
+		FlashWriteError,
+		RomActivationError,
+		OutOfMemoryError,
+		InternalError
+	};
+
+	ErrorCode errorCode = NoError;
+
+	static String errorToString(ErrorCode code);
+
+	size_t write(const uint8_t* data, size_t size) override;
+
+	bool hasError() const
+	{
+		return (state == StateError);
+	}
+	
+	// overrides from IDataSourceStream
+    uint16_t readMemoryBlock(char* data, int bufSize) override
+    {
+        return 0;
+    }
+	virtual int available() override
+	{
+		return 0;
+	}
+    bool isFinished() override
+    {
+        return true;
+    }
+
+private:
+	struct Slot
+	{
+		uint8_t index;
+		uint32_t address;
+		uint32_t size;
+		bool updated = false;
+
+		Slot();
+	} slot;
+
+	// Instead of RbootOutputStream, the rboot write API is used directly because in a future extension the OTA file may contain data for multiple FLASH regions.
+	rboot_write_status rbootWriteStatus = {};
+
+	enum State {
+		StateError,
+		StateHeader,
+		StateRomHeader,
+		StateSkipRom,
+		StateWriteRom,
+		StateRomSignature,
+		StateRomsComplete
+	} state = StateHeader;
+
+	struct {
+		uint32_t magic;
+		uint32_t buildTimestampLow;
+		uint32_t buildTimestampHigh;
+		struct {
+			uint8_t romCount;
+			uint8_t reserved[3];
+		};
+	} fileHeader;
+
+	union {
+		struct {
+			uint32_t address;
+			uint32_t size;
+		} romHeader;
+#ifdef OTA_SIGNED
+		uint8_t signature[crypto_sign_BYTES];
+#else
+		uint8_t signature[16]; // md5sum
+#endif
+	};
+
+	size_t remainingBytes = 0;
+	uint8_t *destinationPtr = nullptr;
+	uint8_t romIndex = 0;
+
+#ifdef OTA_SIGNED
+	crypto_sign_state verifierState;
+#else
+	struct {
+		// see esptool/flasher_stub/rom_functions.h
+		uint32_t state[4];
+  		uint32_t count[2];
+  		uint8_t buffer[64];
+	} md5Context;
+#endif
+
+#ifdef OTA_ENCRYPTED
+	struct {
+		crypto_secretstream_xchacha20poly1305_state state;
+
+		union {
+			uint8_t header[crypto_secretstream_xchacha20poly1305_HEADERBYTES];
+			uint16_t chunkSizeMinusOne;
+		};
+
+		enum {
+			FragmentHeader,
+			FragmentChunkSize,
+			FragmentChunk,
+			FragmentNone
+		} fragment = FragmentHeader;
+		
+		size_t remainingBytes;
+		uint8_t *fragmentPtr;
+		uint8_t *buffer = nullptr;
+		size_t bufferSize = 0;
+	} encryption;
+#endif
+
+	void setError(ErrorCode ec);
+
+	size_t process(const uint8_t* data, size_t size);
+
+	void setupChunk(State nextState, size_t size, void *destination = nullptr)
```

Comment:
Add doc comments to all methods that are not overriding stuff from ReadWriteStream and parents.

Reply:
I've added documentation to the public methods/members. Would you like the private methods to be documented, too?